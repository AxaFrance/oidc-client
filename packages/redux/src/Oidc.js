import React from 'react';
import { OidcProvider, loadUser } from 'redux-oidc';
import { compose, lifecycle } from 'recompose';
import PropTypes from 'prop-types';
import { OidcRoutes, authenticationService, getUserManager, configurationPropTypes } from '@axa-fr/react-oidc-core';
import AuthenticationCallback from './AuthenticationCallback';

const propTypes = {
  notAuthenticated: PropTypes.elementType,
  notAuthorized: PropTypes.elementType,
  // eslint-disable-next-line react/require-default-props
  configuration: configurationPropTypes,
  store: PropTypes.object.isRequired,
  isEnabled: PropTypes.bool,
  children: PropTypes.node,
  callbackComponentOverride: PropTypes.elementType,
};

const defaultPropsObject = {
  notAuthenticated: null,
  notAuthorized: null,
  isEnabled: true,
  children: null,
  callbackComponentOverride: null,
};

const withComponentOverrideProps = (Component, customProps) => props => (
  <Component callbackComponentOverride={customProps} {...props} />
);

export const OidcBase = props => {
  const {
    isEnabled,
    children,
    store,
    callbackComponentOverride,
    configuration,
    notAuthenticated,
    notAuthorized,
  } = props;

  if (!isEnabled) {
    return <>{children}</>;
  }

  return (
    <OidcProvider store={store} userManager={getUserManager()}>
      <OidcRoutes
        configuration={configuration}
        notAuthenticated={notAuthenticated}
        notAuthorized={notAuthorized}
        callbackComponent={withComponentOverrideProps(
          AuthenticationCallback,
          callbackComponentOverride
        )}
      >
        {children}
      </OidcRoutes>
    </OidcProvider>
  );
};

OidcBase.propTypes = propTypes;
OidcBase.defaultProps = defaultPropsObject;

const lifecycleComponent = {
  UNSAFE_componentWillMount() {
    const { isEnabled, store, configuration } = this.props;
    if (isEnabled) {
      const userManager = authenticationService(configuration);
      loadUser(store, userManager);
    }
  },
};

const enhance = compose(lifecycle(lifecycleComponent));

export default enhance(OidcBase);
