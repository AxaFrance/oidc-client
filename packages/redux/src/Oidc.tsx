import { authenticationService, configurationPropTypes, OidcRoutes, UserStoreType } from '@axa-fr/react-oidc-core';
import PropTypes from 'prop-types';
import React, { ComponentType, FC, PropsWithChildren } from 'react';
import { OidcProvider } from 'redux-oidc';
import AuthenticationCallback from './AuthenticationCallback';

const propTypes = {
  notAuthenticated: PropTypes.elementType,
  notAuthorized: PropTypes.elementType,
  callbackComponentOverride: PropTypes.elementType,
  sessionLostComponent: PropTypes.elementType,
  // eslint-disable-next-line react/require-default-props
  configuration: configurationPropTypes,
  store: PropTypes.object.isRequired,
  children: PropTypes.node,
  UserStore: PropTypes.func,
};

const defaultPropsObject: Partial<OidcBaseProps> = {
  notAuthenticated: null,
  notAuthorized: null,
  callbackComponentOverride: null,
  sessionLostComponent: null,
  children: null,
  UserStore: null,
};

const withComponentOverrideProps = (Component: ComponentType, customProps: any) => (props: any) => (
  <Component callbackComponentOverride={customProps} {...props} />
);

export const OidcBaseInternal = (props: any) => {
  const {
    children,
    store,
    callbackComponentOverride,
    configuration,
    notAuthenticated,
    notAuthorized,
    sessionLostComponent,
    UserStore,
    authenticationServiceInternal,
  } = props;

  const getUserManager = () => authenticationServiceInternal(configuration, UserStore);
  return (
    <OidcProvider store={store} userManager={getUserManager()}>
      <OidcRoutes
        configuration={configuration}
        notAuthenticated={notAuthenticated}
        notAuthorized={notAuthorized}
        sessionLost={sessionLostComponent}
        callbackComponent={withComponentOverrideProps(AuthenticationCallback, callbackComponentOverride)}
      >
        {children}
      </OidcRoutes>
    </OidcProvider>
  );
};

OidcBaseInternal.propTypes = {
  ...propTypes,
  authenticationServiceInternal: PropTypes.func.isRequired,
};

type OidcBaseProps = PropsWithChildren<{
  notAuthenticated?: ComponentType | null;
  notAuthorized?: ComponentType | null;
  callbackComponentOverride?: ComponentType | null;
  sessionLostComponent?: ComponentType | null;
  configuration: any;
  store: any;
  UserStore: UserStoreType;
}>;

const OidcBase: FC<OidcBaseProps> = props => <OidcBaseInternal authenticationServiceInternal={authenticationService} {...props} />;

// @ts-ignore
OidcBase.propTypes = propTypes;
OidcBase.defaultProps = defaultPropsObject;

export default OidcBase;
