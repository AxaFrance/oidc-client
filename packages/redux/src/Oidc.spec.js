// Link.react.test.js
import React from "react";
import renderer from "react-test-renderer";

import Oidc from "./Oidc";

// I used __esModule to mock a module with default and named exports
// cf: https://github.com/facebook/jest/issues/5579#issuecomment-397406174
jest.mock("./authenticationService", () => ({
  __esModule: true,
  default: jest.fn(),
  getUserManager: jest.fn()
}));
jest.mock("redux-oidc", () => ({
  OidcProvider: jest.fn(() => <p>Render Something</p>),
  loadUser: jest.fn()
}));

describe("redux.Oidc", () => {
  it("Render <Oidc/> correctly", () => {
    const component = renderer.create(
      <Oidc isEnabled={false} store={{}} configuration={{}}>
        <p>isEnabled</p>
      </Oidc>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
