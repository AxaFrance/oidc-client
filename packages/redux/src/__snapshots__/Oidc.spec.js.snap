// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux-fetch.withAuthentication Render <Oidc/> correctly 1`] = `
<p>
  isEnabled
</p>
`;
